Request batch when idle

akhenry opened the pull request:
<!--- Note: Please open the PR in draft form until you are ready for active review. -->
Closes <!--- Insert Issue Number(s) this PR addresses. Start by typing # will open a dropdown of recent issues. Note: this does not work on PRs which target release branches -->

### Describe your changes:
<!--- Describe your changes and add any comments about your approach either here or inline if code comments aren't added -->

### All Submissions:

* [ ] Have you followed the guidelines in our [Contributing document](https://github.com/nasa/openmct/blob/master/CONTRIBUTING.md)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/nasa/openmct/pulls) for the same update/change?
* [ ] Is this a notable change that will require a special callout in the release notes [Notable Change](../docs/src/process/release.md) ? For example, will this break compatibility with existing APIs or projects which source these plugins?

### Author Checklist

* [ ] Changes address original issue?
* [ ] Tests included and/or updated with changes?
* [ ] Has this been smoke tested?
* [ ] Have you associated this PR with a `type:` label? Note: this is not necessarily the same as the original issue.
* [ ] Have you associated a milestone with this PR? Note: leave blank if unsure.
* [ ] Is this a breaking change to be called out in the release notes?
* [ ] Testing instructions included in associated issue OR is this a dependency/testcase change?

### Reviewer Checklist

* [ ] Changes appear to address issue?
* [ ] Reviewer has tested changes by following the provided instructions?
* [ ] Changes appear not to be breaking changes?
* [ ] Appropriate automated tests included?
* [ ] Code style and in-line documentation are appropriate?

akhenry commented on src/api/telemetry/WebSocketWorker.js:
```diff
@@ -245,7 +245,6 @@ export default function installWorker() {
     }
```
Will try and figure out how to get Webpack to inline Workers so we can do imports etc. Should not block merge right now, it's a no-risk maintenance change we can make later.

akhenry replied:
Some discussion on this here with some potential options - https://github.com/webpack/webpack/discussions/14066

akhenry replied:
This could happen post-merge as it's low risk. It either loads or doesn't.

akhenry commented on src/api/telemetry/WebSocketWorker.js:
```diff
@@ -352,6 +364,33 @@ export default function installWorker() {
     }
   }
 
+  function throttle(callback, wait) {
```
Needed because can't import in workers with current build config.

akhenry commented on src/plugins/openInNewTabAction/openInNewTabAction.js:
```diff
@@ -33,6 +33,6 @@ export default class OpenInNewTab {
   invoke(objectPath, urlParams = undefined) {
     let url = objectPathToUrl(this._openmct, objectPath, urlParams);
-    window.open(url);
+    window.open(url, undefined, "noopener");
```
`noopener` means that the new window/tab does not share the same event loop as the opening window. It makes debugging A LOT easier and probably gives a subtle performance boost.

akhenry replied:
See http://hassansin.github.io/shared-event-loop-among-same-origin-windows

akhenry commented on src/plugins/telemetryTable/components/TableComponent.vue:
```diff
@@ -498,11 +498,17 @@ export default {
   },
   created() {
     this.filterTelemetry = _.debounce(this.filterTelemetry, 500);
+    const throttledFunc = _.throttle(
```
Call throttle callback on leading-edge so that screen update happens in same event loop iteration as telemetry receipt. Improves debugging and also means telemetry is not delayed an additional second beyond the delay created by telemetry batching.

akhenry commented on src/plugins/telemetryTable/components/TableComponent.vue:
```diff
@@ -498,11 +498,17 @@ export default {
   },
   mounted() {
     this.csvExporter = new CSVExporter();
-    this.rowsAdded = _.throttle(this.rowsAdded, 200);
```
Throttling updateVisibleRows now, no need to also throttle `rowsAdded` and `rowsRemoved`

akhenry commented on src/plugins/telemetryTable/components/TableComponent.vue:
```diff
@@ -575,9 +579,12 @@ export default {
         this.updatingView = this.renderWhenVisible(() => {
+          // Cancel any other pending updates to avoid unnecessary callbacks in 1s time.
+          // Otherwise we get superfluous delayed updates that waste CPU cycles.
+          cancelOutstandingInvocations();
```
TODO: can we get rid of this if we do leading-edge throttling _only_ ?

akhenry commented on src/plugins/telemetryTable/components/TableComponent.vue:
```diff
@@ -750,18 +757,18 @@ export default {
       if (!this.sizingRows[sizingRow.objectKeyString]) {
         this.sizingRows[sizingRow.objectKeyString] = sizingRow;
-        this.$nextTick().then(this.calculateColumnWidths);
+        //this.$nextTick().then(this.calculateColumnWidths);
```
TODO: restore this.

akhenry commented on src/plugins/telemetryTable/components/TableComponent.vue:
```diff
@@ -816,18 +823,18 @@ export default {
     },
     calculateTableSize() {
-      this.$nextTick().then(this.calculateColumnWidths);
+      //this.$nextTick().then(this.calculateColumnWidths);
```
TODO: Restore this

akhenry replied:
Done.

akhenry commented on src/plugins/telemetryTable/components/TableComponent.vue:
```diff
@@ -816,18 +823,18 @@ export default {
 
       this.updateHeaders();
-      this.$nextTick().then(this.calculateColumnWidths);
+      //this.$nextTick().then(this.calculateColumnWidths);
```
TODO: Restore this

akhenry replied:
Done.

akhenry commented on src/plugins/telemetryTable/components/TableComponent.vue:
```diff
@@ -816,18 +823,18 @@ export default {
     addObject() {
       this.updateHeaders();
-      this.$nextTick().then(this.calculateColumnWidths);
+      //this.$nextTick().then(this.calculateColumnWidths);
```
TODO: Restore this

akhenry replied:
Done.

akhenry commented on src/plugins/telemetryTable/components/TableRow.vue:
```diff
@@ -180,9 +180,7 @@ export default {
-      return this.columnWidths[key] === undefined
-        ? {}
-        : { width: this.columnWidths[key] + 'px', 'max-width': this.columnWidths[key] + 'px' };
+      return {};
```
TODO: Restore this

akhenry replied:
Done.

akhenry commented on src/plugins/timeConductor/ConductorAxis.vue:
```diff
@@ -92,7 +92,7 @@ export default {
     //Respond to changes in conductor
     this.openmct.time.on(TIME_CONTEXT_EVENTS.timeSystemChanged, this.setViewFromTimeSystem);
-    this.resizeTimer = setInterval(this.resize, RESIZE_POLL_INTERVAL);
+    //this.resizeTimer = setInterval(this.resize, RESIZE_POLL_INTERVAL);
```
TODO: Replace with resize observer

akhenry replied:
Done.

akhenry commented on src/ui/components/TimeSystemAxis.vue:
```diff
@@ -100,7 +100,7 @@ export default {
     this.setDimensions();
     this.drawAxis(this.bounds, this.timeSystem);
-    this.resizeTimer = setInterval(this.resize, RESIZE_POLL_INTERVAL);
+    //this.resizeTimer = setInterval(this.resize, RESIZE_POLL_INTERVAL);
```
TODO: Replace with resize observer

akhenry replied:
Done

akhenry commented on src/plugins/telemetryTable/components/TableComponent.vue:
```diff
@@ -255,8 +255,8 @@
               :style="{
-                width: configuredColumnWidths[key] + 'px',
-                'max-width': configuredColumnWidths[key] + 'px'
+                'width': '30%',
```
Restore dynamic sizing.

akhenry replied:
Done.

shefalijoshi commented on src/api/telemetry/WebSocketWorker.js:
```diff
@@ -335,14 +349,16 @@ export default function installWorker() {
       const batch = this.#batch;
+      batch.number = Math.round(Math.random() * 100000);
       this.#resetBatch();
+      this.sent = Date.now();
```
Doesn't look like `this.sent` is being used. Please remove it.

akhenry replied:
Will do, thanks

shefalijoshi reviewed (changes requested):
This all looks good.

During testing I found one issue - after websocket is disconnected (simulated by killing quickstart locally) and reconnects after a try or two, the remote clock doesn't tick anymore. I don't think this is related to the batching changes though?

<img width="1674" alt="image" src="https://github.com/nasa/openmct/assets/5247066/3992399c-c55e-47b7-b90a-aa5faf3d01c6">

shefalijoshi commented on src/plugins/timeConductor/ConductorAxis.vue:
```diff
@@ -92,7 +91,7 @@ export default {
     //Respond to changes in conductor
     this.openmct.time.on(TIME_CONTEXT_EVENTS.timeSystemChanged, this.setViewFromTimeSystem);
-    this.resizeTimer = setInterval(this.resize, RESIZE_POLL_INTERVAL);
+    //this.resizeTimer = setInterval(this.resize, RESIZE_POLL_INTERVAL);
```
If this is no longer needed, please remove it.

akhenry commented:
> During testing I found one issue - after websocket is disconnected (simulated by killing quickstart locally) and reconnects after a try or two, the remote clock doesn't tick anymore. I don't think this is related to the batching changes though?

Good catch, thanks, lemme take a look.

akhenry commented on src/api/telemetry/WebSocketWorker.js:
```diff
@@ -100,7 +101,7 @@ export default function installWorker() {
     }
 
     #connected() {
```
good question, lemme take a look

akhenry commented:
> This all looks good.
> 
> During testing I found one issue - after websocket is disconnected (simulated by killing quickstart locally) and reconnects after a try or two, the remote clock doesn't tick anymore. I don't think this is related to the batching changes though?

This was a really good catch. There was some functionality missing to re-subscribe after a WebSocket drops, which was a bad miss on my part. 

I've fixed the issue and added a test to catch a regression here.

shefalijoshi reviewed (approved):
Looks good!

akhenry commented:
@shefalijoshi  Addressed issue with `requestIdleCallback`. Thanks again!

shefalijoshi reviewed (approved):
Looks good to me.
